src/_C014.ino

```diff
@@ -0,0 +1,686 @@
+      {
+        byte ControllerID = findFirstEnabledControllerWithId(CPLUGIN_ID_014);
+        if (ControllerID == CONTROLLER_MAX) {
+          // Controller is not enabled.
+          break;
+        } else {
+          String cmd;
+          struct EventStruct TempEvent;
+          TempEvent.TaskIndex = event->TaskIndex;
+          TempEvent.Source = VALUE_SOURCE_MQTT; // to trigger the correct feedback
+          bool validTopic = false;
+          int lastindex = event->String1.lastIndexOf('/');
+          const String lastPartTopic = event->String1.substring(lastindex + 1);
+          errorCounter = 0;
+          log=F("C014 : MQTT received: ");
+          if (event->String1.substring(lastindex + 1) == F("set"))
+          {
+            pubname = event->String1.substring(0,lastindex);
+            lastindex = pubname.lastIndexOf('/');
+            String nodeName = pubname.substring(0,lastindex);
+            String valueName = pubname. substring(lastindex+1);
+            lastindex = nodeName.lastIndexOf('/');
+            nodeName = nodeName. substring(lastindex+1);
+            log+=F("/set: N: ");
+            log+=nodeName;
+            log+=F(" V: ");
+            log+=valueName;
+
+            if (nodeName == F(CPLUGIN_014_SYSTEM_DEVICE)) // msg to a system device
+            {
+              if (valueName.substring(0,strlen(CPLUGIN_014_GPIO_VALUE)) == F(CPLUGIN_014_GPIO_VALUE)) // msg to to set gpio values
+              {
+                cmd = ("GPIO,");
+                cmd += valueName.substring(strlen(CPLUGIN_014_GPIO_VALUE)).toInt(); // get the GPIO
+                if (event->String2=="true" || event->String2=="1") cmd += F(",1");
+                  else cmd += F(",0");
+                validTopic = true;
+              } else if (valueName==CPLUGIN_014_CMD_VALUE) // msg to send a command
+              {
+                cmd = event->String2;
+                validTopic = true;
+              } else
+              {
+                cmd = "SYSTEM not found";
+              }
+            } else if (nodeName == F(CPLUGIN_014_CMD_VALUE)) // legethy homie/system/cmd topic
+            {
+              cmd = event->String2;
+              TempEvent.Source = VALUE_SOURCE_MQTT; // to trigger the correct feedback
+              validTopic = true;
+            } else // msg to a receiving plugin (?)
+            {
+              int deviceNr=CPlugin_014_getPluginNr(nodeName);
+              int pluginID=Device[getDeviceIndex(Settings.TaskDeviceNumber[deviceNr])].Number;
+
+              if (pluginID==12) // Plugin 12 LCD
+              {
+                cmd = valueName; // LCD or LCDCMD
+                cmd += F(",");
+                cmd += event->String2; // line,pos,Text or command as payload
+                validTopic = true;
+              } else if (pluginID==33) // Plugin 33 Dummy Device
+              { // DummyValueSet,<task/device nr>,<value nr>,<value/formula (!ToDo) >, works only with new version of P033!
```

Comment:
I am not a fan of handling plugin commands in a controller.
Why have plugin 12 and 33 to be handled here in the controller and not in other controllers?
Or is there another way to reroute any command from a controller to the normal way of handling commands?
Then we can make a single list of commands that can be handled originating from "outside" the ESPeasy scope as some kind of 'firewall'.
This way we have to look at every controller that handles specific plugin commands.

Reply:
I agree here, my initial concept was and still is to use the homie protocol and it´s autodetect features to make integrating a diy device running ESPEasy as flawless as possible for an end user and experienced user too. If you defined the unit name, all your devices, all their values in ESPEasy it is quite convenient not doing the same in your controller software again trying to match all topics, value names and types ... 
As you wrote before ESPEasy started as a "sensor" firmware sending data to `.../%value`. But when it comes to actors like gpio, rgb leds like neopixel ... there must be a defined route to pass information back. Homie uses a simple` .../%value%/set` path to do that.
Later I found out that the dummy device is a good way to pass numeric data to ESPEasy. This then can be handled by rules and passed to devices as nessesary.
For all commands (or better commands protected by ESPEasy) there is the cmd path as used before (i.e. in the openHAB plugin) only now coming from homie/%unitname%/SYSTEM/cmd/set instead /%unitname%/cmd.

But for some commands (but not many) I think it can be useful to have a direct way from a controller item to a value within ESPEasy. I.e. a switch item to a gpio pin performing ON/OFF action. Therefor the controller plugin not really handling the command it is more **translating** it from its topic and payload to the corresponding command.
The dummy device can be nicely connected to a slider / dimmer to control the pwm duty cycle for a LED or setting other values.
The LCD was only a test device for strings I had sitting next to me, but I will drop this because it makes no sense. It need 3 values (line, position and text), and ON OFF or CLEAR are not realy useful to bbe integrated in an UI without code.

The criteria for this translation should be:

- usefull for user interaction with standard UI elements (boolean, numeric value, RGB(W), HSV, text(?), enum (perhaps for IR controls))
- (none or) only one parameter necessary.
- no security risk.

So there is not much left beside gpio, dummy or neopixel / RGB(W) dimmers to pass data. In the future there could be a way for plugins to advertise their receive value capabilities beside commands. If this is added they should provide the information for the `$format` parameter too:

```
$format | Device → Controller Describes what are valid values for this property:
from:to Describes a range of values e.g. 10:15. Valid for datatypes integer, float (good for sliders and gauges)
value,value,value for enumerating all valid values.  e.g. A,B,C or ON,OFF,PAUSE. Valid for datatypes enum (remote controls)
rgb to provide colors in RGB format e.g. 255,255,0 for yellow. (sorry no RGBW here)
hsv to provide colors in HSV format e.g. 60,100,100 for yellow. Valid for datatype color 
```

As the controller plugin still use the PluginCall() path as the original openHAB plugin does so I see no actual security risk. 

So I`ll  drop the LCD part and keep 

- homie/%unitname%/SYSTEM/cmd/set to pass commands
- homie/%unitname%/SYSTEM/gpio#/set for switches
- homie/%unitname%/%dummyDeviceName%/%valueName%/set for dummy devices

I`ll do some experiments with RGB neopixels and perhaps come up with a solution for port expanders like PCA9685 or PCF8574 (which I use a lot because of the limited GPIOs available). For PCF8574 I do not have a idea how to detect if one port expander is present and how the ports will be used. Doing a i2c scan and rely on the i2c address found is not the right way I think.

Sorry for the long comment but I think the overall concept of using the homie convention should be easy to use user experience  beside standardization. I know that everything is currently already working in some way or another (using rules, mqtt import plugin, rules inside the controler, matching naming and so on) but not in a convenient or consistent way.

Reply:
For the port expanders there is some activity in the background.
In short, I want a general GPIO handler which can be extended to port expanders.
This will then allow other plugins to use these GPIO's too.
So please have a quick search first or open a new issue about that one.

And about the communication in "the other direction". Those will result in a rather big change.
For the inter-ESPeasy communication we have the p2p protocol (part integrated in C013) which I plan to extend further. This can be a start to see what is really needed for the 2-direction communication. I still feel like it should be a task of the controller, but on the other hand we do not want to do all work again for a new controller. So I'm not really sure about that one yet.

Reply:
I understand. I changed the code that currently ONLY gpio and DummyVaueSet are handled via /set and acknowledgements. This works fine, even when the command comes from another source like rule or HTTP. 
But if a different command like LongPulse changes the gpio state there will be no acknowledgement and the controller will be out of sync. 
Therefore the acknowledgement call should be placed directly where gpios are addressed. If there is already some work in progress for 2 way communication we should be happy for now and handle these cases with rules:

```
On gpioSeconds do
  gpio,%eventvalue1%,1
  let,1,%eventvalue1%
  timerSet,1,%eventvalue2%
endon

On Rules#Timer=1 do  //When Timer1 expires, do
  gpio,[VAR#1],0
endon
```